File: 39-CombinationSum/39-CombinationSum.cpp
// Last updated: 7/24/2025, 8:35:55 AM
static const auto speedUp = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();

class Solution {
private:
    vector<vector<int>> res;
    vector<int> current;

    void helper(vector<int>& candidates, int target, int n) {
        if (n < 0 || target < 0)
            return;
        if (target == 0)
            res.push_back(current);
        for (int i = n; i >= 0; i--) {
            current.push_back(candidates[i]);
            helper(candidates, target - candidates[i], i);
            current.pop_back();
        }
    }

public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        helper(candidates, target, candidates.size() - 1);
        return res;
    }
};

File: 39-CombinationSum/39-CombinationSum.py
# Last updated: 7/24/2025, 8:36:01 AM
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = set()

        def helper(current, currentSum):
            for num in candidates:
                nextSum = currentSum + num
                if nextSum == target:
                    res.add(tuple(sorted(current + [num])))
                elif nextSum < target:
                    helper(current + [num], nextSum)

        helper([], 0)
        return [list(combination) for combination in res]
